docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,181 @@
+
+The original proposal reserve block segments for different tiers, we can't do that without modify tendermint, but it should be good enough to just specify higher priority for higher tier transactions. 
+
+The lowest tier can use constant gas price, the higher tiers adjust gas price based on block load at different speed.
+
+##### Positive
+
+- Predictable at the lowest/default tier.
+- Can prioritize urgent transactions in higher tiers.
+- Deterministic at all tiers, no price auction at all.
+
+##### Negative
+
+- TODO
+
+## Decision
+
+After comparing the trade-offs of different designs, we think the adapted multi-tier system is most suitable for cosmos-sdk's case.
+
+We can adopt EIP-1559's approach to adjust higher tier's gas prices based on block load.
+
+### Implementation
+
+We can make each tier's gas price strategy fully configurable in protocol parameters, while providing a sensible default one.
+
+Pseudocode in python-like syntax:
+
+```python
+interface TieredTx:
+  def tier(self) -> int:
+    pass
+
+def tx_tier(tx):
+    if isinstance(tx, TieredTx):
+      return tx.tier()
+    else:
+      # default tier for custom transactions
+      return 0
+
+class TierParams:
+  'gas price strategy parameters of one tier'
+  priority: int           # priority in tendermint mempool
+  initial_gas_price: Coin
+  parent_gas_target: int
+  change_speed: Decimal   # 0 means don't adjust for block load.
+
+class Params:
+    'protocol parameters'
+    tiers: List[TierParams]
+
+class State:
+    'consensus state'
+    # total gas used in last block, None when it's the first block
+    parent_gas_used: Optional[int]
+    # gas prices of last block for all tiers
+    gas_prices: List[Coin]
+
+def begin_block():
+    'Adjust gas prices'
+    for i, tier in enumerate(Params.tiers):
+        if State.parent_gas_used is None:
+            # initialized gas price for the first block
+	          State.gas_prices[i] = tier.initial_gas_price
+        else:
+            gas_used_change_rate = (State.parent_gas_used - tier.parent_gas_target) // tier.parent_gas_target
+            State.gas_prices[i] *= 1 + gas_used_change_rate * tier.change_speed
+
+def mempoolFeeTxHandler_checkTx(ctx, tx):
+    # the minimal-gas-price configured by validator, zero in deliver_tx context
+    validator_price = ctx.MinGasPrice()
+    consensus_price = State.gas_prices[tx_tier(tx)]
+    min_price = max(validator_price, consensus_price)
+
+    if tx.gas_price() < min_price:
+        return 'insufficient fees'
+    return next_CheckTx(ctx, tx)
+
+def txPriorityHandler_checkTx(ctx, tx):
+    res, err := next_CheckTx(ctx, tx)
+    # pass priority to tendermint
+    res.Priority = Params.tiers[tx_tier(tx)].priority
```

Comment:
we still need to order transactions in a tire.  Probably we still need to add gas price component here.

Reply:
I think the point in this design is to not order tx by price directly. Of course we still have the issue if people all flooded to the higher tier, the lower tier won't have a chance to get included, the potential solution reserves block area for each tier, but that'll need to change tendermint too.

Reply:
We know that ordering only by price is not the perfect solution, but it's the most general one which is good for simapp example, until we have something better. We want transactions prioritized by type , price and and transaction distribution (by type).

That being said, an ADR proposal must be more complete.

Reply:
Some of the tires can definitely be full, so we should decide about the order. It can be FIFO or or anything... I think ordering by gas price makes most of the sense in this design.

Reply:
the order would be `FIFO` under the same tier/priority.

Reply:
OK, so let's add it to the doc

Reply:
Tendermint 0.35 can already prioritize transactions, and we also have a primitive example in the SDK how to use it.